spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/SpannerTransactionManager.java

```diff
@@ -0,0 +1,208 @@
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.data.spanner.core;
+
+import javax.annotation.Nullable;
+
+import com.google.cloud.spanner.AbortedException;
+import com.google.cloud.spanner.DatabaseClient;
+import com.google.cloud.spanner.Key;
+import com.google.cloud.spanner.KeySet;
+import com.google.cloud.spanner.Mutation;
+import com.google.cloud.spanner.Options;
+import com.google.cloud.spanner.ResultSet;
+import com.google.cloud.spanner.SpannerException;
+import com.google.cloud.spanner.Statement;
+import com.google.cloud.spanner.Struct;
+import com.google.cloud.spanner.TransactionContext;
+import com.google.cloud.spanner.TransactionManager;
+
+import org.springframework.dao.DuplicateKeyException;
+import org.springframework.transaction.NoTransactionException;
+import org.springframework.transaction.TransactionDefinition;
+import org.springframework.transaction.TransactionException;
+import org.springframework.transaction.UnexpectedRollbackException;
+import org.springframework.transaction.interceptor.TransactionAspectSupport;
+import org.springframework.transaction.support.AbstractPlatformTransactionManager;
+import org.springframework.transaction.support.DefaultTransactionStatus;
+
+/**
+ * Spanner transaction manager
+ *
+ * @author Alexander Khimich
+ */
+public class SpannerTransactionManager extends AbstractPlatformTransactionManager {
+	private final DatabaseClient databaseClient;
+
+	public SpannerTransactionManager(final DatabaseClient databaseClient) {
+		this.databaseClient = databaseClient;
+	}
+
+	protected Tx getCurrentTX() {
+		try {
+			return (SpannerTransactionManager.Tx) ((DefaultTransactionStatus) TransactionAspectSupport
+					.currentTransactionStatus())
+							.getTransaction();
+		}
+		catch (NoTransactionException e) {
+			return null;
+		}
+	}
+
+	@Override
+	protected Object doGetTransaction() throws TransactionException {
+		Tx tx = getCurrentTX();
+		if (tx != null
+				&& tx.transactionManager.getState() == TransactionManager.TransactionState.STARTED) {
+			logger.debug(tx + " reuse; state = " + tx.transactionManager.getState());
+			return tx;
+		}
+		// create a new one if current is not there
+		tx = new Tx();
+		tx.transactionManager = this.databaseClient.transactionManager();
+		logger.debug(tx + " create; state = " + tx.transactionManager.getState());
+		return tx;
+	}
+
+	@Override
+	protected void doBegin(Object transactionObject, TransactionDefinition transactionDefinition)
+			throws TransactionException {
+		Tx tx = (Tx) transactionObject;
+		final TransactionContext targetTransactionContext = tx.transactionManager.begin();
+		if (transactionDefinition.isReadOnly()) {
```

Comment:
It looks like there's a function `DatabaseClient.readOnlyTransaction` can provide a read-only context object here.

Reply:
correct, however that is different class returned

Reply:
Yes, but I think it should be accommodated because of the potentially large performance difference.

Reply:
So, should this be a different issue?